daemon/daemon.go

```diff
@@ -148,28 +148,35 @@ func (daemon *Daemon) Install(eng *engine.Engine) error {
 	return nil
 }
 
-// Get looks for a container by the specified ID or name, and returns it.
-// If the container is not found, or if an error occurs, nil is returned.
-func (daemon *Daemon) Get(name string) *Container {
-	id, err := daemon.idIndex.Get(name)
-	if err == nil {
-		return daemon.containers.Get(id)
-	}
+// Get looks for a container with the provided prefix
+func (daemon *Daemon) Get(prefix string) (*Container, error) {
+	if containerByID := daemon.containers.Get(prefix); containerByID != nil {
 
-	if c, _ := daemon.GetByName(name); c != nil {
-		return c
+		// prefix is an exact match to a full container ID
+		return containerByID, nil
 	}
 
-	if err == truncindex.ErrDuplicateID {
-		log.Errorf("Short ID %s is ambiguous: please retry with more characters or use the full ID.\n", name)
+	containerByName, _ := daemon.GetByName(prefix)
+	containerId, indexError := daemon.idIndex.Get(prefix)
+
+	if containerByName != nil {
```

Comment:
@acbodine is correct, he is implementing what Eric and I discussed, and a few comments after the linked one above I expanded those thoughts with some pseudo-code here: https://gist.github.com/estesp/1e1d3b17d35acf3f9259  and Andy has implemented that here.. I checked over it in an earlier version of the patch to validate it follows that flow, and if we want to double-check I think Andy added a test that validates this as well.